lib/propolis/src/lifecycle.rs

```diff
@@ -0,0 +1,89 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use futures::future::{self, BoxFuture};
+
+use crate::migrate::Migrator;
+
+/// General trait for emulated devices in the system.
+///
+/// As the VM goes through its lifecycle, the emulated devices which it contains
+/// are so driven through those phases via various event functions (`start`,
+/// `pause`, `resume`, etc).
+///
+/// Note: Calling any of these methods on a [Lifecycle] instance should be done
+/// from a context with access to a [Tokio runtime](tokio::runtime::Runtime):
+///
+/// - [Lifecycle::start]
+/// - [Lifecycle::pause]
+/// - [Lifecycle::resume]
+/// - [Lifecycle::paused]
+/// - [Lifecycle::reset]
+/// - [Lifecycle::halt]
+/// - [Lifecycle::migrate]
+pub trait Lifecycle: Send + Sync + 'static {
+    /// Unique name for devices for a given type
+    ///
+    /// Intended to be `const` once stabilized for trait functions.
+    fn type_name(&self) -> &'static str;
```

Comment:
This is an aspect of the `Entity`/`Lifecycle` system I'd like to revisit in the somewhat near future -- possibly for outright removal.

We're working toward a future state where devices can be instantiated directly from migration payloads (rather than created and then hydrated w/ the migration state).  Depending on the details of that, we might nix the `type_name` stuff entirely.

With all that in mind, I'm incline to leave it for now.